pkg/pillar/cmd/volumemgr/handlers.go

```diff
@@ -0,0 +1,640 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package volumemgr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/lf-edge/edge-containers/pkg/registry"
+	zconfig "github.com/lf-edge/eve/api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/cas"
+	"github.com/lf-edge/eve/pkg/pillar/containerd"
+	"github.com/lf-edge/eve/pkg/pillar/diskmetrics"
+	"github.com/lf-edge/eve/pkg/pillar/tgt"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	utils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+	"github.com/lf-edge/eve/pkg/pillar/zfs"
+)
+
+type volumeHandler interface {
+	// prepareVolume to be created
+	prepareVolume() error
+	// checkAndHandleCreated handles preparation done
+	checkAndHandlePrepared() (bool, error)
+	// createVolume handles creation process
+	createVolume() (string, error)
+	// checkAndHandleCreated handles creation done
+	checkAndHandleCreated() (bool, error)
+	// destroyVolume handles destroy process
+	destroyVolume() (string, error)
+	// populate fills VolumeStatus with the information from previously created volume
+	populate() (bool, error)
+	// getVolumeDetails returns ActualSize, VirtualSize, Format, DirtyFlag and possible error
+	getVolumeDetails() (uint64, uint64, string, bool, error)
+	// usageFromStatus returns calculated usage of volume
+	// with current options of VolumeStatus
+	// to be used in storage usage calculation
+	usageFromStatus() uint64
+}
+
+func getVolumeHandler(volumeManagerContext *volumemgrContext, status *types.VolumeStatus) volumeHandler {
+	common := commonVolumeHandler{volumeManagerContext: volumeManagerContext, status: status}
```

Comment:
Does `volumeHandler` becomes owner of `VolumeStatus`? I mean, is it the only writer, while the caller (the rest of volumemgr) is only reading from the status from here on?
If possible, I think it would be safer and more clear to put handlers into a sub-package of volumemgr and export methods, including something to get the current volume status (or just fields/values that handlers are managing/calculating), while leaving attributes unexported. Currently, it is not very clear what is the separation between a volumeHandler and the rest of the volumemgr.

Reply:
Basically something like the separation between `domainmgr` and `hypervisor` packages.

Reply:
Good point, I split out the logic into separate package. Will work on potential complains from yetus.